src/OrchardCore/OrchardCore.Email.Core/EmailDeliveryServiceResolver.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+using OrchardCore.Email.Services;
+
+namespace OrchardCore.Email;
+
+public class EmailDeliveryServiceResolver : IEmailDeliveryServiceResolver
+{
+    private readonly IServiceProvider _serviceProvider;
+
+    public EmailDeliveryServiceResolver(IServiceProvider serviceProvider)
+    {
+        _serviceProvider = serviceProvider;
+    }
+
+    public IEmailDeliveryService Resolve(string name)
+    {
+        var emailDeliveryServiceDictionary = _serviceProvider.GetEmailDeliveryServiceDictionary();
+
+        if (!emailDeliveryServiceDictionary.TryGetValue(name, out var emailDeliveryService))
+        {
+            emailDeliveryService = emailDeliveryServiceDictionary[EmailDeliveryServiceName.None];
+        }
+
+        return emailDeliveryService;
```

Comment:
> return _serviceProvider.GetKeyedService<IEmailDeliveryService>(name) ?? _serviceProvider.GetRequiredService<IEmailDeliveryService>();


How this will work while all the services are registered as keyed services?

Reply:
We might need to return the `NullEmailDeliveryService` as a fallback, right?

Reply:
Now the implementation has become much simpler, to know why it was before, please look into #15100. This is needed when you need to list all the keyed services alongside with their keys

Reply:
I just realized why I did that before because the `deliveryMethodName` is optional with `null` This will allow the caller to pass a delivery method or use the last registered service. Removing the old code will let the key `null` always which implies to use `NullEmailDeliveryService`

Reply:
`_serviceProvider.GetRequiredService<IEmailDeliveryService>()` will return the first registered service, and should work the same for keyed services. You don't need to manually fall back to `NullEmailDeliveryService` because if there's nothing else registered, it'll be returned anyway.

I don't know if a `null` key would work with `GetKeyedService()`, perhaps it wouldn't, in which case indeed you have to add a null check yourself.

Reply:
> _serviceProvider.GetRequiredService<IEmailDeliveryService>() will return the first registered service, and should work the same for keyed services.

The keyed service should need a key in advance, for that we need to get all the keyed services for a particular type, you could check my latest commit

Reply:
The implementation gets even more complex. It seems to me that now there's a re-implementation of keyed services on top of keyed services.

I added a commit to demonstrate what I mean: https://github.com/OrchardCMS/OrchardCore/commit/9d44f4c5f7f31a841cabf5bf4c0efda385280931 This works with or without `deliveryMethodName`, with one or both implementation modules enabled.

Reply:
As I said before `IEmailService` is a key service, if you try what you suggested it will throw an exception. For that I made a keyed dictionary which is missing in ASP.NET Core implementation

Reply:
Another suggestion is to add an `IEmailDeliveryService` implementation if no one exists in DI, otherwise replace the old one, this one will keep track of the latest registered service

Reply:
The code I've shown doesn't cause an exception, I tested it. I don't see a reason to make it any more complex than that.

Reply:
> The code I've shown doesn't cause an exception, I tested it

I tested too, please let me know how the `IEmailDeliveryService` could be retrieved while it's registered as a keyed service.

Reply:
@Piedone this is the main thing that I need to address ASAP to finalize this PR

Reply:
The magic is that it's [also registered as a normal scoped service](https://github.com/OrchardCMS/OrchardCore/commit/9d44f4c5f7f31a841cabf5bf4c0efda385280931#diff-d939f4b3e424bdd1f75122bcd33b41f48e572812526df1a702a68aa848b37de2R22). Since it's resolved either as such or a keyed service, this won't result in a double resolution. Registering these three services (at a given time, at most two) twice doesn't have any practical drawback.

Reply:
Similar to what I suggest [here](https://github.com/OrchardCMS/OrchardCore/pull/14749#discussion_r1474291683), unfortunately I didn't notice your scope registration, I think I will borrow glasses :)

Reply:
I really don't see us benefiting from debating this any further. Please either use my implementation, or provide an alternative, as code in a commit, that's at least as succint or simpler, but not more complex.

Reply:
> I really don't see us benefiting from debating this any further

There's no debating :) it seems you got me wrong. I already did what you suggested

Reply:
Thank you!